File: strip-docx-fonts.js
/**
 * Strip embedded font binaries from .docx before API upload.
 * Word "Embed fonts in the file" can inflate a 1-page letter past Vercel's
 * ~4.5 MB request limit; stripping client-side keeps uploads small.
 */
(function (global) {
	'use strict';

	const EMBED_TAG_RE = /<w:embed(?:Regular|Bold|Italic|BoldItalic)\b[^>]*\/>/gi;
	const FONT_REL_RE = /<Relationship\b[^>]*\bTarget="fonts\/[^"]*"[^>]*\/>/gi;
	const FONT_CONTENT_TYPE_RE = /<Default\b[^>]*\bExtension="(?:odttf|ttf)"[^>]*\/>/gi;
	const FONT_PREFIX = 'word/fonts/';

	function hasEmbeddedFonts(zip) {
		return Object.keys(zip.files).some((name) => name.startsWith(FONT_PREFIX));
	}

	async function prepare(arrayBuffer) {
		if (!arrayBuffer || !global.JSZip) {
			return { bytes: new Uint8Array(arrayBuffer || 0), stripped: false, bytesSaved: 0 };
		}

		const zip = await global.JSZip.loadAsync(arrayBuffer);
		if (!hasEmbeddedFonts(zip)) {
			return { bytes: new Uint8Array(arrayBuffer), stripped: false, bytesSaved: 0 };
		}

		let bytesSaved = 0;
		for (const name of Object.keys(zip.files)) {
			if (name.startsWith(FONT_PREFIX)) {
				const entry = zip.files[name];
				if (entry && !entry.dir) {
					bytesSaved += (await entry.async('uint8array')).length;
				}
				delete zip.files[name];
			}
		}

		const fontTable = zip.file('word/fontTable.xml');
		if (fontTable) {
			const text = await fontTable.async('string');
			zip.file('word/fontTable.xml', text.replace(EMBED_TAG_RE, ''));
		}

		const fontRels = zip.file('word/_rels/fontTable.xml.rels');
		if (fontRels) {
			const text = (await fontRels.async('string')).replace(FONT_REL_RE, '');
			if (text.trim()) {
				zip.file('word/_rels/fontTable.xml.rels', text);
			} else {
				delete zip.files['word/_rels/fontTable.xml.rels'];
			}
		}

		const contentTypes = zip.file('[Content_Types].xml');
		if (contentTypes) {
			const text = await contentTypes.async('string');
			zip.file('[Content_Types].xml', text.replace(FONT_CONTENT_TYPE_RE, ''));
		}

		const bytes = await zip.generateAsync({
			type: 'uint8array',
			compression: 'DEFLATE',
			compressionOptions: { level: 6 },
		});

		return { bytes, stripped: true, bytesSaved };
	}

	function toBase64(bytes) {
		const chunk = 0x8000;
		let binary = '';
		for (let i = 0; i < bytes.length; i += chunk) {
			binary += String.fromCharCode.apply(null, bytes.subarray(i, i + chunk));
		}
		return btoa(binary);
	}

	global.NcStripDocxFonts = { prepare, toBase64 };
})(typeof window !== 'undefined' ? window : globalThis);

File: api/strip_docx_fonts.py
"""
Strip embedded font binaries from .docx uploads.

Word's "Embed fonts in the file" option can inflate a 1-page letter to 5+ MB.
NcFormatter only needs text/structure for IR extraction; embedded fonts are unused
and can push base64 payloads past Vercel's ~4.5 MB request limit.
"""
from __future__ import annotations

import io
import re
import zipfile

_EMBED_TAG_RE = re.compile(
	r"<w:embed(?:Regular|Bold|Italic|BoldItalic)\b[^>]*/>",
	re.IGNORECASE,
)
_FONT_REL_RE = re.compile(
	r'<Relationship\b[^>]*\bTarget="fonts/[^"]*"[^>]*/>',
	re.IGNORECASE,
)
_ODTTF_CONTENT_TYPE_RE = re.compile(
	r'<Default\b[^>]*\bExtension="odttf"[^>]*/>',
	re.IGNORECASE,
)
_FONT_PART_PREFIXES = ("word/fonts/",)


def strip_embedded_docx_fonts(file_bytes: bytes) -> tuple[bytes, bool, int]:
	"""
	Return (docx_bytes, stripped, bytes_saved).

	If no embedded fonts are present, returns the original bytes unchanged.
	"""
	if not file_bytes:
		return file_bytes, False, 0

	font_entries: list[tuple[zipfile.ZipInfo, bytes]] = []
	other_entries: list[tuple[zipfile.ZipInfo, bytes]] = []

	with zipfile.ZipFile(io.BytesIO(file_bytes), "r") as zin:
		for item in zin.infolist():
			data = zin.read(item.filename)
			if item.filename.startswith(_FONT_PART_PREFIXES):
				font_entries.append((item, data))
			else:
				other_entries.append((item, data))

	if not font_entries:
		return file_bytes, False, 0

	saved = sum(len(data) for _, data in font_entries)
	out = io.BytesIO()
	with zipfile.ZipFile(out, "w", compression=zipfile.ZIP_DEFLATED) as zout:
		for item, data in other_entries:
			name = item.filename
			if name == "word/fontTable.xml":
				data = _EMBED_TAG_RE.sub("", data.decode("utf-8")).encode("utf-8")
			elif name == "word/_rels/fontTable.xml.rels":
				text = _FONT_REL_RE.sub("", data.decode("utf-8"))
				if text.strip():
					data = text.encode("utf-8")
				else:
					continue
			elif name == "[Content_Types].xml":
				data = _ODTTF_CONTENT_TYPE_RE.sub("", data.decode("utf-8")).encode("utf-8")
			zout.writestr(item, data)

	return out.getvalue(), True, saved
